pinot-common/src/main/java/org/apache/pinot/common/metrics/PinotMetricUtils.java

```diff
@@ -18,82 +18,186 @@
+import java.util.concurrent.TimeUnit;
 import java.util.function.Function;
-import org.apache.pinot.common.exception.InvalidConfigException;
+import org.apache.pinot.spi.annotations.metrics.MetricsFactory;
+import org.apache.pinot.spi.annotations.metrics.PinotMetricsFactory;
 import org.apache.pinot.spi.env.PinotConfiguration;
 import org.apache.pinot.spi.metrics.PinotGauge;
 import org.apache.pinot.spi.metrics.PinotJmxReporter;
+import org.apache.pinot.spi.metrics.PinotMeter;
 import org.apache.pinot.spi.metrics.PinotMetricName;
 import org.apache.pinot.spi.metrics.PinotMetricsRegistry;
-import org.apache.pinot.common.metrics.yammer.YammerGauge;
-import org.apache.pinot.common.metrics.yammer.YammerJmxReporter;
-import org.apache.pinot.common.metrics.yammer.YammerMetricName;
-import org.apache.pinot.common.metrics.yammer.YammerMetricsRegistry;
+import org.apache.pinot.spi.metrics.PinotTimer;
+import org.apache.pinot.spi.utils.PinotReflectionUtils;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 
 public class PinotMetricUtils {
   private static final Logger LOGGER = LoggerFactory.getLogger(PinotMetricUtils.class);
-  public static final String LIBRARY_NAME_KEY = "libraryName";
-  public static final String YAMMER_KEY = "yammer";
+  private static final String METRICS_PACKAGE_REGEX_PATTERN = ".*\\.plugin\\.metrics\\..*";
 
-  private static String LIBRARY_TO_USE = YAMMER_KEY;
+  private static PinotMetricsFactory _pinotMetricsFactory = null;
 
-  public static void init(PinotConfiguration metricsConfiguration)
-      throws InvalidConfigException {
-    String libraryName = metricsConfiguration.getProperty(PinotMetricUtils.LIBRARY_NAME_KEY);
-    if (libraryName == null) {
-      return;
-    }
-    switch (libraryName) {
-      case YAMMER_KEY:
-        LIBRARY_TO_USE = YAMMER_KEY;
-        break;
-      // TODO: support more libraries.
-      default:
-        throw new InvalidConfigException("PinotMetricsRegistry for " + libraryName + " cannot be initialized.");
+  private static final Map<PinotMetricsRegistry, Boolean> metricsRegistryMap = new ConcurrentHashMap<>();
+  private static final Map<MetricsRegistryRegistrationListener, Boolean> metricsRegistryRegistrationListenersMap =
+      new ConcurrentHashMap<>();
+
+  public static void init(PinotConfiguration metricsConfiguration) {
+    // Initializes PinotMetricsFactory.
+    initializePinotMetricsFactory(metricsConfiguration);
+
+    // Initializes metrics using the metrics configuration.
+    initializeMetrics(metricsConfiguration);
+    registerMetricsRegistry(getPinotMetricsRegistry());
+  }
+
+  /**
+   * Initializes PinotMetricsFactory with metrics configurations.
+   * @param metricsConfiguration The subset of the configuration containing the metrics-related keys
+   */
+  private static void initializePinotMetricsFactory(PinotConfiguration metricsConfiguration) {
+    Set<Class<?>> classes = getPinotMetricsFactoryClasses();
+    for (Class<?> clazz : classes) {
```

Comment:
Can we support multiple MetricsFactory classes?

Reply:
In theory, yes we can support multiple MetricsFactory classes. But I don't think it's necessary for most of the use cases, since usually there is only one metrics system got hook to Pinot cluster. I think it'll make more sense to replace a old metric with a new one, rather than keeping multiple running ones. That's why I only put one static MetricsFactory constant in `PinotMetricUtils` class.

Reply:
Agreed, in this case, can you:
1. add a warning log if there are multiple PinotMetricsFactory Classes
2. add an info log of which PinotMetricsFactory we set for the static MetricsFactory constant.
This will help with running and tests new plugins.

Reply:
We do have an info log in Line 163 when registering metrics factory to `PinotMetricUtils` class.
And sure, will add one warning log if more than one MetricsFactory is needed.